sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/TaskKiller.java

```diff
@@ -1,31 +1,128 @@
+                },
+                killInterval.toMillis(),
+                killInterval.toMillis(),
+                TimeUnit.MILLISECONDS);
+    }
 
-    public TaskKiller(SchedulerDriver driver) {
-        this.driver = driver;
+    public static void setDriver(SchedulerDriver schedulerDriver) {
+        driver = schedulerDriver;
     }
 
-    public void killTask(TaskID taskId) {
+    /**
+     * Calling this method will cause the referenced TaskID to be reliably killed.  An attempt to kill the task will be
+     * immediately made.  Continued periodic attempts will be made to kill the task until Mesos indicates that it has
+     * killed the Task or doesn't recongize the task.
+     *
+     * This is still not a guarantee that a task has been killed.  Mesos may not know about a particular TaskID at any
+     * given time.  See the {@link TaskCleaner} for the rest of the task killing system.
+     * @param taskId the TaskID of the Task to be killed.
+     */
+    public static void killTask(TaskID taskId) {
         // In order to update a podinstance its normal to kill all tasks in a pod.
         // Sometimes a task hasn't been launched ever but it has been recorded for
         // resource reservation footprint reasons, and therefore doesn't have a TaskID yet.
         if (taskId.getValue().isEmpty()) {
-            logger.warn("Attempted to kill empty TaskID.");
+            LOGGER.warn("Attempted to kill empty TaskID.");
             return;
         }
 
-        logger.info("Killing task: {}", taskId.getValue());
-        driver.killTask(taskId);
+        synchronized (lock) {
+            LOGGER.info("Enqueueing kill of task: {}", taskId.getValue());
+            tasksToKill.add(taskId);
+        }
+
+        killTaskInternal(taskId);
+    }
+
+    public static void update(Protos.TaskStatus taskStatus) {
+        // The race here is fine.  If we check for the set containing an element
+        // then try to remove it after it's already gone there's no harm.  In return
+        // for this subtlety we avoid acquiring the lock on potentially many TaskStatus
+        // updates for tasks we don't care about killing.
+        if (isDead(taskStatus) && tasksToKill.contains(taskStatus.getTaskId())) {
+            synchronized (lock) {
+                tasksToKill.remove(taskStatus.getTaskId());
+                LOGGER.info("Completed killing: {}", taskStatus.getTaskId());
+            }
+        }
+    }
+
+    @VisibleForTesting
+    static void killAllTasks() {
+        for (TaskID taskId : tasksToKill) {
+            killTaskInternal(taskId);
+        }
+    }
+
+    private static void killTaskInternal(TaskID taskId) {
+        if (driver != null) {
+            LOGGER.info("Killing task: {}", taskId.getValue());
+            driver.killTask(taskId);
+        } else {
+            LOGGER.warn("Can't kill '{}' driver not yet set.", taskId);
+        }
+    }
+
+    private static boolean isDead(Protos.TaskStatus taskStatus) {
```

Comment:
Upstream into TaskUtils with the other TaskState checks?

Reply:
Maybe.  I've got two of these special ones now.  There's another over in the `TaskCleaner`.  They're a bit service specific and probably shouldn't be used by other components.  I could move them up though.  I'll think about it.